File: v2.x/Examples/SciChartDemo/SciChartDemo/Views/CreateStockCharts/RealtimeTickingStockChart/RealtimeTickingStockChartView.m
//******************************************************************************
// SCICHART® Copyright SciChart Ltd. 2011-2018. All rights reserved.
//
// Web: http://www.scichart.com
// Support: support@scichart.com
// Sales:   sales@scichart.com
//
// RealtimeTickingStockChartView.m is part of the SCICHART® Examples. Permission is hereby granted
// to modify, create derivative works, distribute and publish any part of this source
// code whether for commercial, private or personal use.
//
// The SCICHART® examples are distributed in the hope that they will be useful, but
// without any warranty. It is provided "AS IS" without warranty of any kind, either
// expressed or implied.
//******************************************************************************

#import "RealtimeTickingStockChartView.h"
#import "DataManager.h"
#import "PriceSeries.h"
#import "MovingAverage.h"
#import "MarketDataService.h"
#import "NSDate+missingMethods.h"

static int const DefaultPointCount = 150;
static uint const SmaSeriesColor = 0xFFFFA500;
static uint const StrokeUpColor = 0xFF00AA00;
static uint const StrokeDownColor = 0xFFFF0000;

@implementation RealtimeTickingStockChartView {
    uint _smaSeriesColor;
    uint _strokeUpColor;
    uint _strokeDownColor;

    SCIOhlcDataSeries * _ohlcDataSeries;
    SCIXyDataSeries * _xyDataSeries;
    
    SCIAxisMarkerAnnotation * _smaAxisMarker;
    SCIAxisMarkerAnnotation * _ohlcAxisMarker;
    
    MarketDataService * _marketDataService;
    MovingAverage * _sma50;
    PriceBar * _lastPrice;
    
    PriceUpdateCallback onNewPriceBlock;
}

- (void)commonInit {
    __weak typeof(self) wSelf = self;
    onNewPriceBlock = ^(PriceBar * price) { [wSelf onNewPrice:price]; };
    self.continueTickingTouched = ^{ [wSelf subscribePriceUpdate]; };
    self.pauseTickingTouched = ^{ [wSelf clearSubscribtions]; };
    self.seriesTypeTouched = ^{ [wSelf changeSeriesType]; };
}

- (void)initExample {
    _marketDataService = [[MarketDataService alloc] initWithStartDate:[NSDate dateWithYear:2000 month:8 day:01 hour:12 minute:0 second:0] TimeFrameMinutes:5 TickTimerIntervals:0.02];
    _sma50 = [[MovingAverage alloc] initWithLength:50];
    
    [self initDataWithService:_marketDataService];

    [self createMainPriceChart];
    SCIBoxAnnotation * leftAreaAnnotation = [SCIBoxAnnotation new];
    SCIBoxAnnotation * rightAreaAnnotation = [SCIBoxAnnotation new];
    
    [self createOverviewChartWithLeftAnnotation:leftAreaAnnotation RightAnnotation:rightAreaAnnotation];
    
    SCIAxisBase * axis = (SCIAxisBase *)[self.mainSurface.xAxes itemAt:0];
    [axis registerVisibleRangeChangedCallback:^(id<SCIRangeProtocol> newRange, id<SCIRangeProtocol> oldRange, BOOL isAnimated, id sender) {
        leftAreaAnnotation.x1 = [self.overviewSurface.xAxes itemAt:0].visibleRange.min;
        leftAreaAnnotation.x2 = [self.mainSurface.xAxes itemAt:0].visibleRange.min;
        
        rightAreaAnnotation.x1 = [self.mainSurface.xAxes itemAt:0].visibleRange.max;
        rightAreaAnnotation.x2 = [self.overviewSurface.xAxes itemAt:0].visibleRange.max;
    }];

    [_marketDataService subscribePriceUpdate:onNewPriceBlock];
}

- (void)initDataWithService:(MarketDataService *)marketDataService {
    _ohlcDataSeries = [[SCIOhlcDataSeries alloc] initWithXType:SCIDataType_DateTime YType:SCIDataType_Double];
    _ohlcDataSeries.seriesName = @"Price Series";
    _xyDataSeries = [[SCIXyDataSeries alloc] initWithXType:SCIDataType_DateTime YType:SCIDataType_Double];
    _xyDataSeries.seriesName = @"50-Period SMA";
    
    PriceSeries * prices = [marketDataService getHistoricalData:DefaultPointCount];
    _lastPrice = [prices lastObject];

    [_ohlcDataSeries appendRangeX:SCIGeneric(prices.dateData)
                             Open:SCIGeneric(prices.openData)
                             High:SCIGeneric(prices.highData)
                              Low:SCIGeneric(prices.lowData)
                            Close:SCIGeneric(prices.closeData)
                            Count:prices.size];
    double movingAverageArray[prices.size];
    for (int i = 0; i < prices.size; i++) {
        movingAverageArray[i] = [_sma50 push:[prices itemAt:i].close].current;
    }
    [_xyDataSeries appendRangeX:SCIGeneric(prices.dateData) Y:SCIGeneric((double *)movingAverageArray) Count:prices.size];
}

- (void)createMainPriceChart {
    id<SCIAxis2DProtocol> xAxis = [SCICategoryDateTimeAxis new];
    xAxis.growBy = [[SCIDoubleRange alloc] initWithMin:SCIGeneric(0.0) Max:SCIGeneric(0.1)];
    xAxis.style.drawMajorGridLines = NO;
    
    id<SCIAxis2DProtocol> yAxis = [SCINumericAxis new];
    yAxis.autoRange = SCIAutoRange_Always;
    
    SCIFastOhlcRenderableSeries * ohlcSeries = [SCIFastOhlcRenderableSeries new];
    ohlcSeries.dataSeries = _ohlcDataSeries;

    SCIFastLineRenderableSeries * ma50Series = [SCIFastLineRenderableSeries new];
    ma50Series.dataSeries = _xyDataSeries;
    ma50Series.strokeStyle = [[SCISolidPenStyle alloc] initWithColorCode:0xFFFF6600 withThickness:1];
    
    _smaAxisMarker = [SCIAxisMarkerAnnotation new];
    _smaAxisMarker.position = SCIGeneric(0);
    _smaAxisMarker.style.backgroundColor = [UIColor fromARGBColorCode:SmaSeriesColor];
    
    _ohlcAxisMarker = [SCIAxisMarkerAnnotation new];
    _ohlcAxisMarker.position = SCIGeneric(0);
    _ohlcAxisMarker.style.backgroundColor = [UIColor fromARGBColorCode:StrokeUpColor];
    
    SCIZoomPanModifier * zoomPanModifier = [SCIZoomPanModifier new];
    zoomPanModifier.direction = SCIDirection2D_XDirection;
    
    SCILegendModifier * legendModifier = [SCILegendModifier new];
    legendModifier.orientation = SCIOrientationHorizontal;
    legendModifier.position = SCILegendPositionBottom;
    
    [SCIUpdateSuspender usingWithSuspendable:self.mainSurface withBlock:^{
        [self.mainSurface.xAxes add:xAxis];
        [self.mainSurface.yAxes add:yAxis];
        [self.mainSurface.renderableSeries add:ma50Series];
        [self.mainSurface.renderableSeries add:ohlcSeries];
        [self.mainSurface.annotations add:_smaAxisMarker];
        [self.mainSurface.annotations add:_ohlcAxisMarker];
        
        self.mainSurface.chartModifiers = [[SCIChartModifierCollection alloc] initWithChildModifiers:@[[SCIXAxisDragModifier new], zoomPanModifier, [SCIZoomExtentsModifier new], legendModifier]];
        [SCIThemeManager applyDefaultThemeToThemeable:self.mainSurface];
    }];
}

- (void)createOverviewChartWithLeftAnnotation:(SCIBoxAnnotation *)leftAreaAnnotation RightAnnotation:(SCIBoxAnnotation *)rightAreaAnnotation {
    id<SCIAxis2DProtocol> xAxis = [SCICategoryDateTimeAxis new];
    xAxis.autoRange = SCIAutoRange_Always;
    
    id<SCIAxis2DProtocol> yAxis = [SCINumericAxis new];
    yAxis.growBy = [[SCIDoubleRange alloc] initWithMin:SCIGeneric(0.1) Max:SCIGeneric(0.1)];
    yAxis.autoRange = SCIAutoRange_Always;
    
    SCIFastMountainRenderableSeries * mountainSeries = [SCIFastMountainRenderableSeries new];
    mountainSeries.dataSeries = _ohlcDataSeries;
    mountainSeries.areaStyle = [[SCILinearGradientBrushStyle alloc] initWithColorCodeStart:0x883a668f finish:0xff20384f direction:SCILinearGradientDirection_Vertical];
    
    leftAreaAnnotation.coordinateMode = SCIAnnotationCoordinate_RelativeY;
    leftAreaAnnotation.y1 = SCIGeneric(0);
    leftAreaAnnotation.y2 = SCIGeneric(1);
    leftAreaAnnotation.style.fillBrush = [[SCISolidBrushStyle alloc] initWithColorCode:0x33FFFFFF];
    
    rightAreaAnnotation.coordinateMode = SCIAnnotationCoordinate_RelativeY;
    rightAreaAnnotation.y1 = SCIGeneric(0);
    rightAreaAnnotation.y2 = SCIGeneric(1);
    rightAreaAnnotation.style.fillBrush = [[SCISolidBrushStyle alloc] initWithColorCode:0x33FFFFFF];
    
    [SCIUpdateSuspender usingWithSuspendable:self.overviewSurface withBlock:^{
        [self.overviewSurface.xAxes add:xAxis];
        [self.overviewSurface.yAxes add:yAxis];
        [self.overviewSurface.renderableSeries add:mountainSeries];
        [self.overviewSurface.annotations add:leftAreaAnnotation];
        [self.overviewSurface.annotations add:rightAreaAnnotation];
        
        [SCIThemeManager applyDefaultThemeToThemeable:self.overviewSurface];
    }];
}

- (void)onNewPrice:(PriceBar *)price {
    double smaLastValue;
    if (_lastPrice.date == price.date) {
        [_ohlcDataSeries updateAt:_ohlcDataSeries.count - 1 Open:SCIGeneric(price.open) High:SCIGeneric(price.high) Low:SCIGeneric(price.low) Close:SCIGeneric(price.close)];
        
        smaLastValue = [_sma50 update:price.close].current;
        [_xyDataSeries updateAt:_xyDataSeries.count - 1 Y:SCIGeneric(smaLastValue)];
    } else {
        [_ohlcDataSeries appendX:SCIGeneric(price.date) Open:SCIGeneric(price.open) High:SCIGeneric(price.high) Low:SCIGeneric(price.low) Close:SCIGeneric(price.close)];
        
        smaLastValue = [_sma50 push:price.close].current;
        [_xyDataSeries appendX:SCIGeneric(price.date) Y:SCIGeneric(smaLastValue)];
 
        id<SCIRangeProtocol> visibleRange = self.mainSurface.xAxes[0].visibleRange;
        if (visibleRange.maxAsDouble > _ohlcDataSeries.count) {
            [visibleRange setMinTo:SCIGeneric(visibleRange.minAsDouble + 1) MaxTo:SCIGeneric(visibleRange.maxAsDouble + 1)];
        }
    }
    
    _ohlcAxisMarker.style.backgroundColor = price.close >= price.open ? [UIColor fromARGBColorCode:StrokeUpColor] : [UIColor fromARGBColorCode:StrokeDownColor];
    _ohlcAxisMarker.position = SCIGeneric(price.close);
    _smaAxisMarker.position = SCIGeneric(smaLastValue);
    
    _lastPrice = price;
}

- (void)subscribePriceUpdate {
    [_marketDataService subscribePriceUpdate:onNewPriceBlock];
}

- (void)clearSubscribtions {
    [_marketDataService clearSubscriptions];
}

- (void)changeSeriesType {
    UIAlertController * alertController = [UIAlertController alertControllerWithTitle:@"Series type" message:@"Select series type for the top scichart surface" preferredStyle:UIAlertControllerStyleActionSheet];
    [alertController addAction:[UIAlertAction actionWithTitle:@"CandlestickRenderableSeries" style:UIAlertActionStyleDefault handler:^(UIAlertAction * action) {
        [self changeSeries:[SCIFastCandlestickRenderableSeries new]];
    }]];
    
    [alertController addAction:[UIAlertAction actionWithTitle:@"OhlcRenderableSeries" style:UIAlertActionStyleDefault handler:^(UIAlertAction * action) {
        [self changeSeries:[SCIFastOhlcRenderableSeries new]];
    }]];
    
    [alertController addAction:[UIAlertAction actionWithTitle:@"MountainRenderableSeries" style:UIAlertActionStyleDefault handler:^(UIAlertAction * action) {
        [self changeSeries:[SCIFastMountainRenderableSeries new]];
    }]];
    
    UIViewController * topVC = UIApplication.sharedApplication.delegate.window.rootViewController;
    while (topVC.presentedViewController) {
        topVC = topVC.presentedViewController;
    }
    [topVC presentViewController:alertController animated:YES completion:nil];
}

- (void)changeSeries:(SCIRenderableSeriesBase *)rSeries {
    rSeries.dataSeries = _ohlcDataSeries;
    
    [SCIUpdateSuspender usingWithSuspendable:self.mainSurface withBlock:^{
        [self.mainSurface.renderableSeries removeAt:1];
        [self.mainSurface.renderableSeries add:rSeries];
        [SCIThemeManager applyDefaultThemeToThemeable:rSeries];
    }];
}

- (void)willMoveToWindow:(UIWindow *)newWindow {
    [super willMoveToWindow:newWindow];
    
    if (newWindow == nil) {
        [_marketDataService clearSubscriptions];
    }
}
                                                         
@end

File: v2.x/Examples/SciChartSwiftDemo/SciChartSwiftDemo/Views/CreateStockCharts/RealtimeTickingStockChart/RealtimeTickingStockChartView.swift
//******************************************************************************
// SCICHART® Copyright SciChart Ltd. 2011-2018. All rights reserved.
//
// Web: http://www.scichart.com
// Support: support@scichart.com
// Sales:   sales@scichart.com
//
// RealtimeTickingStockChartView.swift is part of the SCICHART® Examples. Permission is hereby granted
// to modify, create derivative works, distribute and publish any part of this source
// code whether for commercial, private or personal use.
//
// The SCICHART® examples are distributed in the hope that they will be useful, but
// without any warranty. It is provided "AS IS" without warranty of any kind, either
// expressed or implied.
//******************************************************************************

let DefaultPointCount: Int32 = 150
let SmaSeriesColor: uint = 0xFFFFA500
let StrokeUpColor: uint = 0xFF00AA00
let StrokeDownColor: uint = 0xFFFF0000

class RealtimeTickingStockChartView: RealtimeTickingStockChartLayout {
    
    var _smaSeriesColor: uint = 0
    var _strokeUpColor: uint = 0
    var _strokeDownColor: uint = 0
    
    let _ohlcDataSeries = SCIOhlcDataSeries(xType: .dateTime, yType: .double)
    let _xyDataSeries = SCIXyDataSeries(xType: .dateTime, yType: .double)
    
    let _smaAxisMarker = SCIAxisMarkerAnnotation()
    let _ohlcAxisMarker = SCIAxisMarkerAnnotation()
    
    let _marketDataService = MarketDataService(start: NSDate(year: 2000, month: 8, day: 01, hour: 12, minute: 0, second: 0) as Date?, timeFrameMinutes: 5, tickTimerIntervals: 0.02)
    let _sma50 = MovingAverage(length: 50)
    var _lastPrice: PriceBar?
    
    var onNewPriceBlock: PriceUpdateCallback?
    
    override func commonInit() {
        weak var wSelf = self
        onNewPriceBlock = { price in wSelf?.onNewPrice(price!) }
        self.continueTickingTouched = { wSelf?.subscribePriceUpdate() }
        self.pauseTickingTouched = { wSelf?.clearSubscribtions() }
        self.seriesTypeTouched = { wSelf?.changeSeriesType() }
    }

    override func initExample() {
        initDataWithService(_marketDataService!)
        
        createMainPriceChart()
        let leftAreaAnnotation = SCIBoxAnnotation()
        let rightAreaAnnotation = SCIBoxAnnotation()
        
        createOverviewChartWith(leftAreaAnnotation, rightAreaAnnotation: rightAreaAnnotation)
        
        let axis = mainSurface.xAxes[0] as! SCIAxisBase
        axis.registerVisibleRangeChangedCallback { (newRange, oldRange, isAnimated, sender) in
            leftAreaAnnotation.x1 = self.overviewSurface.xAxes[0].visibleRange.min
            leftAreaAnnotation.x2 = self.mainSurface.xAxes[0].visibleRange.min
            
            rightAreaAnnotation.x1 = self.mainSurface.xAxes[0].visibleRange.max
            rightAreaAnnotation.x2 = self.overviewSurface.xAxes[0].visibleRange.max
        }
        
        _marketDataService?.subscribePriceUpdate(onNewPriceBlock)
    }
    
    fileprivate func initDataWithService(_ marketDataService: MarketDataService) {
        _ohlcDataSeries.seriesName = "Price Series"
        _xyDataSeries.seriesName = "50-Period SMA";

        let prices = marketDataService.getHistoricalData(DefaultPointCount)
        let size = prices!.size()
        
        _lastPrice = (prices?.lastObject())!
        
        _ohlcDataSeries.appendRangeX(DataManager.getGenericDataArray(prices!.dateData()),
                                     open: DataManager.getGenericDataArray(prices!.openData()),
                                     high: DataManager.getGenericDataArray(prices!.highData()),
                                     low: DataManager.getGenericDataArray(prices!.lowData()),
                                     close: DataManager.getGenericDataArray(prices!.closeData()),
                                     count: size)
        var movingAverageArray = [Double](repeating: 0, count: Int(size))
        let movingAveragePointer: UnsafeMutablePointer<Double> = UnsafeMutablePointer(mutating: movingAverageArray)
        for i in 0..<size {
            movingAverageArray[Int(i)] = (_sma50?.push(prices!.item(at: i).close).current())!
        }
        _xyDataSeries.appendRangeX(DataManager.getGenericDataArray(prices!.dateData()), y: DataManager.getGenericDataArray(movingAveragePointer), count: size)
    }
    
    fileprivate func createMainPriceChart() {
        let xAxis = SCICategoryDateTimeAxis()
        xAxis.growBy = SCIDoubleRange(min: SCIGeneric(0.0), max: SCIGeneric(0.1))
        xAxis.style.drawMajorGridLines = false
        
        let yAxis = SCINumericAxis()
        yAxis.autoRange = .always
        
        let ohlcSeries = SCIFastOhlcRenderableSeries()
        ohlcSeries.dataSeries = _ohlcDataSeries
        
        let ma50Series = SCIFastLineRenderableSeries()
        ma50Series.dataSeries = _xyDataSeries
        ma50Series.strokeStyle = SCISolidPenStyle(colorCode: 0xFFFF6600, withThickness: 1)
        
        _smaAxisMarker.position = SCIGeneric(0)
        _smaAxisMarker.style.backgroundColor = UIColor.fromARGBColorCode(SmaSeriesColor)
        
        _ohlcAxisMarker.position = SCIGeneric(0)
        _ohlcAxisMarker.style.backgroundColor = UIColor.fromARGBColorCode(StrokeUpColor)
        
        let zoomPanModifier = SCIZoomPanModifier()
        zoomPanModifier.direction = .xDirection
        
        let legendModifier = SCILegendModifier()
        legendModifier.orientation = .horizontal
        legendModifier.position = .bottom
        
        SCIUpdateSuspender.usingWithSuspendable(mainSurface) {
            self.mainSurface.xAxes.add(xAxis)
            self.mainSurface.yAxes.add(yAxis)
            self.mainSurface.renderableSeries.add(ma50Series)
            self.mainSurface.renderableSeries.add(ohlcSeries)
            self.mainSurface.annotations.add(self._smaAxisMarker)
            self.mainSurface.annotations.add(self._ohlcAxisMarker)
            
            self.mainSurface.chartModifiers = SCIChartModifierCollection(childModifiers: [SCIXAxisDragModifier(), zoomPanModifier, SCIZoomExtentsModifier(), legendModifier])
            SCIThemeManager.applyDefaultTheme(toThemeable: self.mainSurface)
        }
    }
    
    fileprivate func createOverviewChartWith(_ leftAreaAnnotation: SCIBoxAnnotation, rightAreaAnnotation: SCIBoxAnnotation) {
        let xAxis = SCICategoryDateTimeAxis()
        xAxis.autoRange = .always
        
        let yAxis = SCINumericAxis()
        yAxis.growBy = SCIDoubleRange(min: SCIGeneric(0.1), max: SCIGeneric(0.1))
        yAxis.autoRange = .always
        
        let mountainSeries = SCIFastMountainRenderableSeries()
        mountainSeries.dataSeries = _ohlcDataSeries
        mountainSeries.areaStyle = SCILinearGradientBrushStyle(colorCodeStart: 0x883a668f, finish: 0xff20384f, direction: .vertical)
        
        leftAreaAnnotation.coordinateMode = .relativeY
        leftAreaAnnotation.y1 = SCIGeneric(0)
        leftAreaAnnotation.y2 = SCIGeneric(1)
        leftAreaAnnotation.style.fillBrush = SCISolidBrushStyle(colorCode: 0x33FFFFFF)
        
        rightAreaAnnotation.coordinateMode = .relativeY
        rightAreaAnnotation.y1 = SCIGeneric(0)
        rightAreaAnnotation.y2 = SCIGeneric(1)
        rightAreaAnnotation.style.fillBrush = SCISolidBrushStyle(colorCode: 0x33FFFFFF)
        
        SCIUpdateSuspender.usingWithSuspendable(overviewSurface) {
            self.overviewSurface.xAxes.add(xAxis)
            self.overviewSurface.yAxes.add(yAxis)
            self.overviewSurface.renderableSeries.add(mountainSeries)
            self.overviewSurface.annotations.add(leftAreaAnnotation)
            self.overviewSurface.annotations.add(rightAreaAnnotation)
            
            SCIThemeManager.applyDefaultTheme(toThemeable: self.overviewSurface)
        }
    }
    
    fileprivate func onNewPrice(_ price: PriceBar) {
        let smaLastValue: Double
        if (_lastPrice!.date == price.date) {
            _ohlcDataSeries.update(at: _ohlcDataSeries.count() - 1, open: SCIGeneric(price.open), high: SCIGeneric(price.high), low: SCIGeneric(price.low), close: SCIGeneric(price.close))
            
            smaLastValue = (_sma50?.update(price.close)?.current())!
            _xyDataSeries.update(at: _xyDataSeries.count() - 1, y: SCIGeneric(smaLastValue))
        } else {
            _ohlcDataSeries.appendX(SCIGeneric(price.date), open: SCIGeneric(price.open), high: SCIGeneric(price.high), low: SCIGeneric(price.low), close: SCIGeneric(price.close))

            smaLastValue = (_sma50?.push(price.close)?.current())!
            _xyDataSeries.appendX(SCIGeneric(price.date), y: SCIGeneric(smaLastValue))
            
            let visibleRange = mainSurface.xAxes[0].visibleRange!
            if (visibleRange.maxAsDouble() > Double(_ohlcDataSeries.count())) {
                visibleRange.setMinTo(SCIGeneric(visibleRange.minAsDouble() + 1), maxTo: SCIGeneric(visibleRange.maxAsDouble() + 1))
            }
        }
        
        _ohlcAxisMarker.style.backgroundColor = price.close >= price.open ? UIColor.fromARGBColorCode(StrokeUpColor) : UIColor.fromARGBColorCode(StrokeDownColor)
        _ohlcAxisMarker.position = SCIGeneric(price.close);
        _smaAxisMarker.position = SCIGeneric(smaLastValue);
        
        _lastPrice = price;
    }
    
    fileprivate func subscribePriceUpdate() {
        _marketDataService?.subscribePriceUpdate(onNewPriceBlock)
    }
    
    fileprivate func clearSubscribtions() {
        _marketDataService?.clearSubscriptions()
    }
    
    fileprivate func changeSeriesType() {
        let alertController = UIAlertController(title: "Series type", message: "Select series type for the top scichart surface", preferredStyle: .actionSheet)
        alertController.addAction(UIAlertAction(title: "CandlestickRenderableSeries", style: .default, handler: {(action: UIAlertAction) -> Void in
            self.changeSeries(SCIFastCandlestickRenderableSeries())
        }))
        alertController.addAction(UIAlertAction(title: "OhlcRenderableSeries", style: .default, handler: {(action: UIAlertAction) -> Void in
            self.changeSeries(SCIFastOhlcRenderableSeries())
        }))
        alertController.addAction(UIAlertAction(title: "MountainRenderableSeries", style: .default, handler: {(action: UIAlertAction) -> Void in
            self.changeSeries(SCIFastMountainRenderableSeries())
        }))
        
        var topVC = UIApplication.shared.delegate?.window??.rootViewController
        while ((topVC?.presentedViewController) != nil) {
            topVC = topVC?.presentedViewController
        }
        topVC?.present(alertController, animated: true, completion: nil)
    }
    
    fileprivate func changeSeries(_ rSeries: SCIRenderableSeriesBase) {
        rSeries.dataSeries = _ohlcDataSeries
        
        SCIUpdateSuspender.usingWithSuspendable(mainSurface) {
            self.mainSurface.renderableSeries.remove(at: 1)
            self.mainSurface.renderableSeries.add(rSeries)
            SCIThemeManager.applyDefaultTheme(toThemeable: rSeries)
        }
    }
    
    override func willMove(toWindow newWindow: UIWindow?) {
        super.willMove(toWindow: newWindow);
        if newWindow == nil {
            _marketDataService?.clearSubscriptions()
        }
    }
}
